File: Week_04/322/coin_change2.py
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        cache = {0: 0}

        def dfs(n: int):
            if n in cache:
                return cache[n]
            if n < 0:
                return -1
            v = amount + 1
            for c in coins:
                if n >= c and dfs(n - c) >= 0:
                    v = min(dfs(n - c) + 1, v)
            cache[n] = v
            return v

        if dfs(amount) == amount + 1:
            return -1
        return dfs(amount)

File: Week_04/322/coin_change2.go
package main

import (
	"math"
)

func coinChange2(coins []int, amount int) int {
	var dfs func(n int) int
	cache := make(map[int]int, 0)
	dfs = func(n int) int {
		if n < 0 {
			return -1
		}
		if n == 0 {
			return 0
		}
		if v, ok := cache[n]; ok {
			return v
		}
		res := math.MaxInt64
		for _, c := range coins {
			if v := dfs(n - c); v >= 0 && v < res {
				res = v + 1
			}
		}
		cache[n] = res
		return res
	}
	if dfs(amount) == math.MaxInt64 {
		return -1
	}
	return dfs(amount)
}
